applidium/pericles
# Improve security schemes

The current security schemes were not completely spec-compliant of Open API.
It answered a quick issue with a part of the spec only.

We take more time to be more spec compliant now.

This commit considers:

* https://swagger.io/specification/#securitySchemeObject
* https://swagger.io/docs/specification/authentication/

This commit does the following:

* Add parameters into security_scheme according to spec
* Rename `parameters` to `specification_extensions` to have the same name that in the spec
* Change the list of security schemes to be simpler
* Change the form of security schemes to consider different cases and attributes
* Add controls into model to ensure spec-compliance
* Correct swagger generation

# QA

<details>
  <summary>New list</summary>

![image](https://user-images.githubusercontent.com/11132978/68018749-3708d680-fc9a-11e9-8d98-add23dca5275.png)
</details>

<details>
  <summary>New form</summary>

![image](https://user-images.githubusercontent.com/11132978/68019146-476d8100-fc9b-11e9-8974-294fa4ebe679.png)

![image](https://user-images.githubusercontent.com/11132978/68019186-56ecca00-fc9b-11e9-8619-7c97b5bd065a.png)

![image](https://user-images.githubusercontent.com/11132978/68019209-623ff580-fc9b-11e9-81e1-2f1d4aa431e4.png)

![image](https://user-images.githubusercontent.com/11132978/68019228-6b30c700-fc9b-11e9-9476-9b6b2cf91221.png)

</details>

<details>
  <summary>Corrected swagger</summary>

Copy/paste into https://editor.swagger.io/ to see the result.
:information_source: Responses in my project are not set.

```json
{
  "openapi": "3.0.0",
  "info": {
    "description": "URL\r\n** Please fill URL PROD if it exists or any other urls relative to this project\r\n\r\nFormat\r\n** Specify JSON / Restful format : https://i.imgur.com/rzoYLSX.png\r\n\r\nEncoding\r\n** Define encoding - utf-8 par défaut : https://i.imgur.com/j7pvEy1.png\r\n\r\nCompression\r\n** Define compression : https://i.imgur.com/Eix5nng.png\r\n\r\nProtocol\r\n** Make call with HTTPS (cf. Apple) : https://i.imgur.com/2a6mD2B.png\r\n\r\nLocalization\r\n** Define localization (header preferably or parameter) : https://i.imgur.com/NFDbaiL.png\r\n\r\nDates\r\n** Our date format should fullfill iso 8601 : https://fr.wikipedia.org/wiki/ISO_8601\r\n\r\nAuthentication\r\n** Define the format of the request and the authentication response =\u003e Example: authentication OAuth =\u003e Think to create a route!\r\n** Define errors and associated messages\r\n** Define the headers for each public / private route\r\n** Define the expiration logic of tokens\r\n",
    "title": "tst",
    "version": "1.0.0"
  },
  "servers": [],
  "tags": [
    {
      "name": "Report"
    }
  ],
  "paths": {
    "/reports": {
      "get": {
        "tags": [
          "Report"
        ]
      }
    },
    "/reports/{id}": {
      "get": {
        "tags": [
          "Report"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ]
      }
    },
    "/reports/{id}/detail": {
      "get": {
        "tags": [
          "Report"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ]
      }
    },
    "/reports/{id}/detail/{subid}": {
      "put": {
        "tags": [
          "Report"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          },
          {
            "name": "subid",
            "in": "path",
            "required": true,
            "schema": {
              "type": "string"
            }
          }
        ]
      }
    }
  },
  "components": {
    "schemas": {
      "DefaultReport": {
        "title": "Report - DefaultReport",
        "type": "object",
        "properties": {
          "id": {
            "type": "string"
          },
          "code": {
            "type": "string"
          }
        },
        "additionalProperties": false,
        "description": "Automatically generated (please edit me)",
        "required": [
          "code",
          "id"
        ]
      }
    },
    "securitySchemes": {
      "withOpenIdConnect": {
        "type": "openIdConnect",
        "description": "This is the description of the ~Security scheme~.",
        "openIdConnectUrl": "http://example.com"
      },
      "withOauth2": {
        "type": "oauth2",
        "description": "This is the description of the _Security scheme_.",
        "flows": {
          "authorizationCode": {
            "scopes": {
              "read": "Grants read access",
              "admin": "Grants access to admin operations",
              "write": "Grants write access"
            },
            "tokenUrl": "https://example.com/oauth/token",
            "authorizationUrl": "https://example.com/oauth/authorize"
          }
        }
      },
      "withHttp": {
        "type": "http",
        "description": "This is the description of the **Security scheme**.",
        "scheme": "basic"
      },
      "withApiKey": {
        "type": "apiKey",
        "description": "This is the description of the *Security scheme*.",
        "in": "header",
        "name": "Authorization"
      }
    }
  }
}
```
</details>
